clone/cmd/ctclone/ctclone.go

```diff
@@ -0,0 +1,178 @@
+// Copyright 2021 Google LLC. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// ctclone is a one-shot tool for downloading entries from a CT log.
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"flag"
+	"fmt"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/google/trillian-examples/clone/internal/download"
+	"github.com/google/trillian-examples/clone/internal/logdb"
+
+	_ "github.com/go-sql-driver/mysql"
+)
+
+const getEntriesFormat = "ct/v1/get-entries?start=%d&end=%d"
+
+var (
+	logURL         = flag.String("log_url", "", "Log storage root URL, https://ct.googleapis.com/rocketeer/")
+	mysqlURI       = flag.String("mysql_uri", "", "URL of a MySQL database to clone the log into. The DB should contain only one log.")
+	fetchBatchSize = flag.Uint("fetch_batch_size", 32, "The number of entries to fetch from the log in each request.")
+	writeBatchSize = flag.Uint("write_batch_size", 32, "The number of leaves to write in each DB transaction.")
+	workers        = flag.Uint("workers", 2, "The number of worker threads to run in parallel to fetch entries.")
+)
+
+func main() {
+	flag.Parse()
+
+	if !strings.HasSuffix(*logURL, "/") {
+		glog.Exit("'log_url' must end with '/'")
+	}
+	if len(*mysqlURI) == 0 {
+		glog.Exit("Missing required parameter 'mysql_uri'")
+	}
+	lu, err := url.Parse(*logURL)
+	if err != nil {
+		glog.Exitf("log_url is invalid: %v", err)
+	}
+
+	ctx := context.Background()
+	db, err := logdb.NewDatabase(*mysqlURI)
+	if err != nil {
+		glog.Exitf("Failed to connect to database: %q", err)
+	}
+	// TODO(mhutchinson): Persist log URI and check here to prevent user-error writing different logs to same DB.
+	head, err := db.Head()
+	if err != nil {
+		if err == logdb.ErrNoDataFound {
+			glog.Infof("failed to find head of database, assuming empty and starting from scratch: %v", err)
```

Comment:
super nit: **F**ailed to ...

Reply:
urgh, I really hate that capitalization is different between direct log messages and errors (especially as error text basically just exists to get dumped to a log anyway). Anyway, done :-)